RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterECLCC.h

```diff
@@ -0,0 +1,297 @@
+  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+  Authors: Jayadharini Jaiganesh and Martin Burtscher
+
+  URL: The latest version of this code is available at
+  https://userweb.cs.txstate.edu/~burtscher/research/ECL-CC/.
+
+  Publication: This work is described in detail in the following paper.
+  Jayadharini Jaiganesh and Martin Burtscher. A High-Performance Connected
+  Components Implementation for GPUs. Proceedings of the 2018 ACM International
+  Symposium on High-Performance Parallel and Distributed Computing, pp. 92-104.
+  June 2018.
+*/
+
+/*
+ The code is modified for the specific use-case of generating topological clusters
+ for PFClustering. It is adapted to work with the Alpaka portability library.
+ 
+*/
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  /* intermediate pointer jumping */
+
+  ALPAKA_FN_ACC inline int representative(const int idx,
+                                          reco::PFClusteringVarsDeviceCollection::View pfClusteringVars) {
+    int curr = pfClusteringVars[idx].pfrh_topoId();
+    if (curr != idx) {
+      int next, prev = idx;
+      while (curr > (next = pfClusteringVars[curr].pfrh_topoId())) {
+        pfClusteringVars[prev].pfrh_topoId() = next;
+        prev = curr;
+        curr = next;
+      }
+    }
+    return curr;
+  }
+
+  // Initial step of ECL-CC. Uses ID of first neighbour in edgeList with a smaller ID
+  class ECLCCInit {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusteringEdgeVarsDeviceCollection::View pfClusteringEdgeVars) const {
+      const int nRH = pfRecHits.size();
+      const int from = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] +
+                       alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u] *
+                           alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
+      for (int v : cms::alpakatools::elements_with_stride(acc, nRH)) {
+        const int beg = pfClusteringEdgeVars[v].pfrh_edgeIdx();
+        const int end = pfClusteringEdgeVars[v + 1].pfrh_edgeIdx();
+        int m = v;
+        int i = beg;
+        while ((m == v) && (i < end)) {
+          m = std::min(m, pfClusteringEdgeVars[i].pfrh_edgeList());
+          i++;
+        }
+        pfClusteringVars[v].pfrh_topoId() = m;
+      }
+
+      if (from == 0) {
```

Comment:
General question: would it be useful to have something like `cms::alpakatools::once_per_kernel(acc)` to simplify this ?

As in
```suggestion
      if ([from == 0](cms::alpakatools::once_per_kernel(acc))) {
```
and removing the definition of `from` above.

Reply:
Possibly? I know it is a common operation in CUDA to either perform an operation using the first thread in a block or the first global thread.

Reply:
`cms::alpakatools::once_per_grid(acc)` is available in https://github.com/cms-sw/cmssw/pull/43205 .